python/cuml/cuml/internals/outputs.py

```diff
@@ -0,0 +1,439 @@
+    >>> model = cuml.DBSCAN(eps=1.0, min_samples=1).fit(X)
+    >>> isinstance(model.labels_, cp.ndarray)
+    True
+
+    With a global output type set though, the fitted attributes will match
+    the configured output type.
+
+    >>> cuml.set_global_output_type("cudf")
+    >>> isinstance(model.labels_, cudf.Series)
+    True
+
+    Reset the output type back to its original value.
+
+    >>> cuml.set_global_output_type(original_output_type)
+    """
+    if output_type is not None:
+        output_type = check_output_type(output_type)
+    GlobalSettings().output_type = output_type
+
+
+class using_output_type:
+    """Configure the output type within a context.
+
+    Parameters
+    ----------
+    output_type : {'input', 'cupy', 'numpy', 'cudf', 'pandas', None}
+        Desired output type of results and attributes of the estimators.
+
+        * ``None``: No globally configured output type. This is the same as
+          ``'input'``, except in cases where an estimator explicitly sets
+          an ``output_type``.
+
+        * ``'input'``: returns arrays of the same type as the inputs to the
+          function or method. Fitted attributes will be of the same array type
+          as ``X``.
+
+        * ``'cupy'``: returns ``cupy`` arrays.
+
+        * ``'numpy'``: returns ``numpy`` arrays.
+
+        * ``'cudf'``: returns ``cudf.Series`` for single dimensional results
+          and ``cudf.DataFrame`` otherwise.
+
+        * ``'pandas'``: returns ``pandas.Series`` for single dimensional results
+          and ``pandas.DataFrame`` otherwise.
+
+    See Also
+    --------
+    cuml.set_global_output_type
+
+    Examples
+    --------
+    >>> import cuml
+    >>> import cupy as cp
+    >>> import cudf
+
+    Fit a model with a cupy array. By default the fitted attributes will be
+    cupy arrays.
+
+    >>> X = cp.array([[1.0, 4.0, 4.0], [2.0, 2.0, 2.0], [5.0, 1.0, 1.0]])
+    >>> model = cuml.DBSCAN(eps=1.0, min_samples=1).fit(X)
+    >>> isinstance(model.labels_, cp.ndarray)
+    True
+
+    With a global output type set though, the fitted attributes will match
+    the configured output type.
+
+    >>> with cuml.using_output_type("cudf"):
+    ...     print(isinstance(model.labels_, cudf.Series))
+    True
+    """
+
+    def __init__(self, output_type):
+        self.output_type = output_type
+
+    def __enter__(self):
+        self.prev_output_type = GlobalSettings().output_type
+        set_global_output_type(self.output_type)
+        return self.prev_output_type
+
+    def __exit__(self, *_):
+        GlobalSettings().output_type = self.prev_output_type
+
+
+@contextlib.contextmanager
+def enter_internal_api():
+    """Enter an internal API context.
+
+    Returns ``True`` if this is a new internal context, or ``False``
+    if the code was already running within an internal context."""
+    gs = GlobalSettings()
+    if gs._external_output_type is False:
```

Comment:
Why not?
```suggestion
    if gs._external_output_type is None:
```

Reply:
`output_type` can be `None`, so we needed another sentinel.

Responding to your other comment here as well - I don't like a custom sentinel for this case since it'd need to be shared (and would probably need to live within `global_settings.py` to avoid import cycles). A value of `False` is easy and works.

I commented the state in `global_settings.py` and hope that's sufficient.

Reply:
Works well for me. I actually thought that I had deleted this comment, because it cannot be `None` for obvious reasons.